### PR TITLE
[UNR-2818] Pin runtime version

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Commands to spatial from the GDK for Unreal now explicitely specify a runtime version to use. A default runtime is chosen by the GDK version, and settings allow the user to change that.
image

#### Documentation
WIP

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h [hunk 1]
```diff
@@ -259,7 +260,22 @@ class SPATIALGDKEDITOR_API USpatialGDKEditorSettings : public UObject
...
 	bool bGenerateDefaultLaunchConfig;
 
+	/** Returns the Runtime version pinned to this release of the GDK. */
+	static const FString& GetPinnedGDKRuntimeVersion();
+
+	/** Returns the Runtime version to use, either the pinned one, or the user-specified one depending of the settings. */
+	const FString& GetSpatialOSRuntimeVersion() const;
+
+	/** Whether to use the GDK-associated SpatialOS runtime version, or to use the one specified in the RuntimeVersion field. */
+	UPROPERTY(EditAnywhere, config, Category = "Runtime")
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h]:
```suggestion
	UPROPERTY(EditAnywhere, config, Category = "Runtime", meta = (DisplayName = "Use GDK Pinned Runtime Version"))
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 2]
```diff
@@ -34,6 +35,23 @@ USpatialGDKEditorSettings::USpatialGDKEditorSettings(const FObjectInitializer& O
 	SpatialOSSnapshotToLoad = GetSpatialOSSnapshotToLoad();
 }
 
+const FString& USpatialGDKEditorSettings::GetPinnedGDKRuntimeVersion()
+{
+	static const FString PinnedVersion("11-20200205T105018Z-7668e9b");
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: Wonder if we should instead stick this inside the `SpatialGDKServices` Module as this is the only module that can be used by all other GDK modules.

**Comment by tenevdev** [comment]
Might be worth adding an option to use latest alongside specifying a version string for better user experience when not using the pinned version.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp [hunk 3]
```diff
@@ -58,6 +58,7 @@ void FSpatialGDKEditorModule::RegisterSettings()
 	}
 
 	FPropertyEditorModule& PropertyModule = FModuleManager::LoadModuleChecked<FPropertyEditorModule>("PropertyEditor");
+
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp]: Redundant newline?

File: SpatialGDK/Source/SpatialGDKServices/Public/SpatialGDKServicesConstants.h [hunk 4]
```diff
@@ -28,5 +28,6 @@ namespace SpatialGDKServicesConstants
 	const FString SpatialExe = CreateExePath(SpatialPath, TEXT("spatial"));
 	const FString SpotExe = CreateExePath(GDKProgramPath, TEXT("spot"));
 	const FString SchemaCompilerExe = CreateExePath(GDKProgramPath, TEXT("schema_compiler"));
-    const FString SpatialOSDirectory = FPaths::ConvertRelativePathToFull(FPaths::Combine(FPaths::ProjectDir(), TEXT("/../spatial/")));
+  const FString SpatialOSDirectory = FPaths::ConvertRelativePathToFull(FPaths::Combine(FPaths::ProjectDir(), TEXT("/../spatial/")));
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDKServices/Public/SpatialGDKServicesConstants.h]: Spacing seems off

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/DeploymentLauncher/DeploymentLauncher.cs [hunk 5]
```diff
@@ -97,10 +97,11 @@ private static int CreateDeployment(string[] args)
 
             var projectName = args[1];
             var assemblyName = args[2];
-            var mainDeploymentName = args[3];
-            var mainDeploymentJsonPath = args[4];
-            var mainDeploymentSnapshotPath = args[5];
-            var mainDeploymentRegion = args[6];
+            var runtimeVersion = args[3];
```
Comment (m-samiec) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/DeploymentLauncher/DeploymentLauncher.cs]: Please increment RequireSetup

File: SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h [hunk 6]
```diff
@@ -259,7 +260,26 @@ class SPATIALGDKEDITOR_API USpatialGDKEditorSettings : public UObject
...
 
+	/** Returns the Runtime version to use for cloud deployments, either the pinned one, or the user-specified one depending of the settings. */
+	const FString& GetSpatialOSRuntimeVersionForCloud() const;
+
+	/** Returns the Runtime version to use for local deployments, either the pinned one, or the user-specified one depending of the settings. */
+	const FString& GetSpatialOSRuntimeVersionForLocal() const;
+
+	/** Whether to use the GDK-associated SpatialOS runtime version, or to use the one specified in the RuntimeVersion field. */
+	UPROPERTY(EditAnywhere, config, Category = "Runtime", meta = (DisplayName = "Use GDK pinned runtime version"))
+	bool bUseGDKPinnedRuntimeVersion;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h]: nit: these all feel like `advanced` options and I'd recommend marking them so

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h]: What do you mean by marking as "advanced" settings ? Putting them somewhere else ? Advanced mention in comments ?
That is also something that user will have to eventually deal with the moment they use the CLI from outside Unreal, so we might want to make it standard and visible.

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h]: Configs can be tagged as `advanced` so they aren't shown by default.

**Pull request merged**